src/local-mode.ts

```diff
@@ -56,51 +102,150 @@ export class LocalModeManager {
+			.then(resp => _.map(resp.Volumes, 'Name'));
+		const networksPromise = this.docker
+			.listNetworks()
+			.then(resp => _.map(resp, 'Id'));
+
+		const data = await Bluebird.all([
+			containersPromise,
+			imagesPromise,
+			volumesPromise,
+			networksPromise,
+		]);
+		return new EngineSnapshotRecord(
+			new EngineSnapshot(data[0], data[1], data[2], data[3]),
+			new Date(),
+		);
 	}
 
-	private async getLocalModeImages(): Promise<string[]> {
-		// Return all local mode images present on the local docker daemon
-		return _.map(
-			await this.docker.listImages({
-				filters: { label: ['io.resin.local.image=1'] },
-			}),
-			'Id',
+	// Store engine snapshot data in the local database.
+	public async storeEngineSnapshot(record: EngineSnapshotRecord) {
+		const timestamp = record.timestamp.toISOString();
+		log.debug(
+			`Storing engine snapshot in the database. Timestamp: ${timestamp}`,
 		);
+		await this.db.models('engineSnapshot').delete();
+		return this.db.models('engineSnapshot').insert({
+			snapshot: JSON.stringify(record.snapshot),
+			timestamp,
+		});
 	}
 
-	private async getLocalModeContainers(
-		localModeImageIds: string[],
-	): Promise<string[]> {
-		return _(await this.docker.listContainers())
-			.filter(({ Image }) => _.includes(localModeImageIds, Image))
-			.map('Id')
-			.value();
+	// Read the latest stored snapshot from the database.
+	public async retrieveLatestSnapshot(): Promise<EngineSnapshotRecord | null> {
+		const r = await this.db
+			.models('engineSnapshot')
+			.select()
+			.orderBy('rowid', 'DESC')
+			.first();
+
+		if (!r) {
+			return null;
+		}
+		return new EngineSnapshotRecord(
+			EngineSnapshot.fromJSON(r.snapshot),
+			new Date(Date.parse(r.timestamp)),
+		);
+	}
+
+	private async removeLocalModeArtifacts(objects: EngineSnapshot) {
+		log.debug(`Going to delete the following objects: ${objects}`);
+
+		// Delete engine objects. We catch every deletion error, so that we can attempt other objects deletions.
+		await Bluebird.map(objects.containers, cId => {
+			return this.docker
+				.getContainer(cId)
+				.remove({ force: true })
+				.catch(e => log.error(`Unable to delete container ${cId}`, e));
+		});
+		await Bluebird.map(objects.images, iId => {
+			return this.docker
+				.getImage(iId)
+				.remove({ force: true })
+				.catch(e => log.error(`Unable to delete image ${iId}`, e));
+		});
+		await Bluebird.map(objects.networks, nId => {
+			return this.docker
+				.getNetwork(nId)
+				.remove()
+				.catch(e => log.error(`Unable to delete network ${nId}`, e));
+		});
+		await Bluebird.map(objects.volumes, vId => {
+			return this.docker
+				.getVolume(vId)
+				.remove()
+				.catch(e => log.error(`Unable to delete volume ${vId}`, e));
+		});
+
+		// Remove any local mode state added to the database.
+		await this.db
+			.models('app')
+			.del()
+			.where({ source: 'local' })
+			.catch(e =>
+				log.error('Cannot delete local app entries in the database', e),
+			);
```

Comment:
Is this new behavior?

Reply:
No, it existed before - just carrying over.
https://github.com/balena-io/balena-supervisor/blob/master/src/local-mode.ts#L77

Reply:
Fair enough, no worries then, I had forgotten about that.